framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,50 +533,82 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	if (!pcm->running) {
```

Comment:
Let's think about role of each API
I think if some code part is for initialzie pcm before record or write, it should be moved to pcm_start.
let's clean readi & playi including only record & play part.

Reply:
pcm_start used to enqueue buffers in record case and then issue ioctl start. 
Now in readi , it is enqueueing the required for recording and then call pcm_start.
I think all enqueue's and deque's should be done in readi and writei.
enqueueing buffers is a part of record, in my opinion.

Reply:
ok I will accept your opinion.